### PR TITLE
UCT/CMA/KNEM: Add errh support

### PR DESCRIPTION
**Pull request by brminich**
## What
Add support for handling peer failures to CMA and KNEM transports

## Why ?
Error handling support would allow  client-server application to use these transports

### CODE REVIEW

File: test/gtest/uct/test_peer_failure.cc [hunk 1]
```diff
@@ -612,3 +614,91 @@ UCS_TEST_SKIP_COND_P(test_uct_peer_failure_keepalive, killed,
...
+
+    uct_rkey_t failed_peer_rkey(mapped_buffer *recv_buffer)
+    {
+        if (has_cma()) {
+            return 0ul;
+        } else {
+            EXPECT_TRUE(has_knem());
+            uct_knem_key_t *rk = (uct_knem_key_t *)recv_buffer->rkey();
+            m_knem_key         = *rk;
+            m_knem_key.cookie = 0;
```
Comment (hoopoepg) [inline test/gtest/uct/test_peer_failure.cc:649]: minor: unaligned ```=```

**Review by hoopoepg — APPROVED** [review]

File: test/gtest/uct/test_peer_failure.cc [hunk 2]
```diff
@@ -612,3 +614,91 @@ UCS_TEST_SKIP_COND_P(test_uct_peer_failure_keepalive, killed,
...
+    uct_knem_key_t   m_knem_key;
+};
+
+UCS_TEST_SKIP_COND_P(test_ucp_peer_failure_rma_zcopy, put,
+                     !check_caps(UCT_IFACE_FLAG_ERRHANDLE_PEER_FAILURE |
+                                 UCT_IFACE_FLAG_PUT_ZCOPY))
+{
+    {
+        scoped_log_handler slh(wrap_errors_logger);
+        const size_t size =128;
```
Comment (dmitrygx) [inline test/gtest/uct/test_peer_failure.cc:664]: minor: add whitespace after `=`

File: test/gtest/uct/test_peer_failure.cc [hunk 3]
```diff
@@ -612,3 +614,91 @@ UCS_TEST_SKIP_COND_P(test_uct_peer_failure_keepalive, killed,
...
+    EXPECT_GT(m_err_count, 0ul);
+}
+
+UCS_TEST_SKIP_COND_P(test_ucp_peer_failure_rma_zcopy, get,
+                     !check_caps(UCT_IFACE_FLAG_ERRHANDLE_PEER_FAILURE |
+                                 UCT_IFACE_FLAG_GET_ZCOPY))
+{
+    {
+        scoped_log_handler slh(wrap_errors_logger);
+        const size_t size =128;
```
Comment (dmitrygx) [inline test/gtest/uct/test_peer_failure.cc:686]: minor: add whitespace after `=`

File: test/gtest/uct/test_peer_failure.cc [hunk 4]
```diff
@@ -612,3 +614,91 @@ UCS_TEST_SKIP_COND_P(test_uct_peer_failure_keepalive, killed,
...
+                     !check_caps(UCT_IFACE_FLAG_ERRHANDLE_PEER_FAILURE |
+                                 UCT_IFACE_FLAG_PUT_ZCOPY))
+{
+    {
+        scoped_log_handler slh(wrap_errors_logger);
+        const size_t size =128;
+        mapped_buffer sendbuf(size, 0, *m_sender);
+        mapped_buffer recvbuf(size, 0, *m_receivers.front());
+
+        ucs_status_t status = uct_ep_put_zcopy(m_sender->ep(0), sendbuf.iov(), 1,
```
Comment (dmitrygx) [inline test/gtest/uct/test_peer_failure.cc:668]: can put/get flows be tested as variants?

File: test/gtest/uct/test_peer_failure.cc [hunk 5]
```diff
@@ -612,3 +614,91 @@ UCS_TEST_SKIP_COND_P(test_uct_peer_failure_keepalive, killed,
...
+        m_dummy_comp.count  = INT_MAX;
+        m_dummy_comp.status = UCS_OK;
+    }
+
+    // Pretend peer failure by using invalid address for cma and invalid key
+    // for knem
+    uint64_t failed_peer_address(mapped_buffer *recv_buffer)
+    {
+        if (has_cma()) {
+            return  0xDEADBEAF;
```
Comment (dmitrygx) [inline test/gtest/uct/test_peer_failure.cc:634]: minor: there is an excessive between `return` and `0xDEADBEAF`

File: test/gtest/uct/test_peer_failure.cc (hunk 4 above)
Comment (brminich) [reply test/gtest/uct/test_peer_failure.cc:670]: could be done this way, but it seems to be simple enough now, so I'd leave it as it is

Comment (brminich) [reply test/gtest/uct/test_peer_failure.cc:670]: also variants for UCT tests were recently removed, so could not do that anyway

Comment (dmitrygx) [reply test/gtest/uct/test_peer_failure.cc:670]: I see, thanks
probably, it's worth wrapping them by to a common wrapper which will take `uct_ep_put_zcopy`/`uct_ep_get_zcopy` as an argument and execute the operation, wdyt?

File: src/uct/sm/scopy/base/scopy_ep.c [hunk 6]
```diff
@@ -163,6 +163,9 @@ ucs_arbiter_cb_result_t uct_scopy_ep_progress_tx(ucs_arbiter_t *arbiter,
             if (tx->iov_iter.iov_index < tx->iov_cnt) {
                 return UCS_ARBITER_CB_RESULT_RESCHED_GROUP;
             }
+        } else {
+            uct_iface_handle_ep_err(&iface->super.super.super, &ep->super.super,
+                                    UCS_ERR_ENDPOINT_TIMEOUT);
```
Comment (yosefe) [inline src/uct/sm/scopy/base/scopy_ep.c:168]: CONNECTION_RESET

File: test/gtest/uct/test_peer_failure.cc [hunk 7]
```diff
@@ -612,3 +614,93 @@ UCS_TEST_SKIP_COND_P(test_uct_peer_failure_keepalive, killed,
...
+    {
+        if (has_cma()) {
+            return 0xDEADBEAF;
+        } else {
+            EXPECT_TRUE(has_knem());
+            return recv_buffer->addr();
+        }
+    }
+
+    uct_rkey_t failed_peer_rkey(mapped_buffer *recv_buffer)
```
Comment (yosefe) [inline test/gtest/uct/test_peer_failure.cc:643]: minor: mapped_buffer&

File: test/gtest/uct/test_peer_failure.cc [hunk 8]
```diff
@@ -612,3 +614,93 @@ UCS_TEST_SKIP_COND_P(test_uct_peer_failure_keepalive, killed,
...
+    test_ucp_peer_failure_rma_zcopy()
+    {
+        m_dummy_comp.func   = NULL;
+        m_dummy_comp.count  = INT_MAX;
+        m_dummy_comp.status = UCS_OK;
+        m_knem_key.cookie   = 0;
+        m_knem_key.address  = 0;
+    }
+
+    // Pretend peer failure by using invalid address for cma and invalid key
```
Comment (yosefe) [inline test/gtest/uct/test_peer_failure.cc:631]: for cma it should be invalid PID

File: test/gtest/uct/test_peer_failure.cc [hunk 9]
```diff
@@ -612,3 +614,93 @@ UCS_TEST_SKIP_COND_P(test_uct_peer_failure_keepalive, killed,
...
+
+    uct_rkey_t failed_peer_rkey(mapped_buffer *recv_buffer)
+    {
+        if (has_cma()) {
+            return 0ul;
+        } else {
+            EXPECT_TRUE(has_knem());
+            uct_knem_key_t *rk = (uct_knem_key_t *)recv_buffer->rkey();
+            m_knem_key         = *rk;
+            m_knem_key.cookie  = 0;
```
Comment (yosefe) [inline test/gtest/uct/test_peer_failure.cc:651]: should be a created and deleted cookie, not just 0

**Review by yosefe — COMMENTED** [review]
in general, i think the test should create and destroy receiver, to emulate error handling, rather than just using valid key/address

**Comment by brminich** [comment]
@hoopoepg, @dmitrygx please review

**Review by hoopoepg — APPROVED** [review]

File: src/uct/base/uct_iface.c [hunk 10]
```diff
@@ -706,9 +708,18 @@ uct_keepalive_info_t* uct_ep_keepalive_create(pid_t pid, ucs_time_t start_time)
...
     }
 
-    ka->start_time = start_time;
     uct_ep_get_process_proc_dir(ka->proc, proc_len + 1, pid);
 
+    status = ucs_sys_get_file_time(ka->proc, UCS_SYS_FILE_TIME_CTIME,
+                                   &start_time);
+    if (status != UCS_OK) {
+        ucs_error("failed to get process start time");
+        ucs_free(ka);
```
Comment (yosefe) [inline src/uct/base/uct_iface.c:717]: maybe use goto for err flow

File: src/uct/sm/scopy/cma/cma_ep.c [hunk 11]
```diff
@@ -37,18 +37,23 @@ const struct {
...
-    UCT_CHECK_PARAM(params->field_mask & UCT_EP_PARAM_FIELD_IFACE_ADDR,
-                    "UCT_EP_PARAM_FIELD_IFACE_ADDR and UCT_EP_PARAM_FIELD_DEV_ADDR are not defined");
-
+    UCT_EP_PARAMS_CHECK_DEV_IFACE_ADDRS(params);
     UCS_CLASS_CALL_SUPER_INIT(uct_scopy_ep_t, params);
+
     self->remote_pid = *(const pid_t*)params->iface_addr &
                        ~UCT_CMA_IFACE_ADDR_FLAG_PID_NS;
+
+    self->keepalive = uct_ep_keepalive_create(self->remote_pid);
```
Comment (yosefe) [inline src/uct/sm/scopy/cma/cma_ep.c:46]: maybe create it on-demand, when ep_check is called. like for mm_ep

File: src/uct/sm/scopy/cma/cma_ep.c [hunk 12]
```diff
@@ -82,9 +87,12 @@ ucs_status_t uct_cma_ep_tx(uct_ep_h tl_ep, const uct_iov_t *iov, size_t iov_cnt,
                                   local_iov_cnt - local_iov_idx, &remote_iov,
                                   1, 0);
     if (ucs_unlikely(ret < 0)) {
-        ucs_error("%s(pid=%d length=%zu) returned %zd: %m",
+        ucs_debug("%s(pid=%d length=%zu) returned %zd: %m",
```
Comment (yosefe) [inline src/uct/sm/scopy/cma/cma_ep.c:90]: log level should be based on return value from uct_iface_handle_ep_err

File: test/gtest/uct/uct_test.cc [hunk 13]
```diff
@@ -623,6 +623,10 @@ bool uct_test::has_cma() const {
     return has_transport("cma");
 }
 
+bool uct_test::has_knem() const {
```
Comment (yosefe) [inline test/gtest/uct/uct_test.cc:626]: why needed?

File: src/uct/sm/scopy/cma/cma_ep.c (hunk 11 above)
Comment (brminich) [reply src/uct/sm/scopy/cma/cma_ep.c:46]: can do that. Should I update cuda_ipc as well? (i recall a discussion where we decided to initialize it always for cuda_ipc)

File: test/gtest/uct/uct_test.cc (hunk 13 above)
Comment (brminich) [reply test/gtest/uct/uct_test.cc:626]: leftover

File: src/uct/sm/scopy/cma/cma_ep.c (hunk 11 above)
Comment (yosefe) [reply src/uct/sm/scopy/cma/cma_ep.c:46]: if can be done easily, then yes

File: src/uct/base/uct_iface.c [hunk 14]
```diff
@@ -714,31 +714,42 @@ uct_keepalive_info_t* uct_ep_keepalive_create(pid_t pid)
...
-                                   &create_time);
-    if (ucs_unlikely((status != UCS_OK) || (ka->start_time != create_time))) {
-        return uct_iface_handle_ep_err(tl_ep->iface, tl_ep,
-                                       UCS_ERR_ENDPOINT_TIMEOUT);
+    if (ucs_unlikely(*ka == NULL)) {
+        *ka = uct_ep_keepalive_create(pid);
+        if (*ka == NULL) {
+            return uct_iface_handle_ep_err(tl_ep->iface, tl_ep,
+                                           UCS_ERR_NO_MEMORY);
+        }
```
Comment (yosefe) [inline src/uct/base/uct_iface.c:744]: what if remote pid is already dead? need to check it

Comment (brminich) [reply src/uct/base/uct_iface.c:744]: `uct_ep_keepalive_create` would fail then

Comment (brminich) [reply src/uct/base/uct_iface.c:744]: @yosefe do you agree with that?

Comment (yosefe) [reply src/uct/base/uct_iface.c:744]: i see, so why is the error UCS_ERR_NO_MEMORY?

Comment (brminich) [reply src/uct/base/uct_iface.c:746]: fixed

**Review by yosefe — APPROVED** [review]

**Review by yosefe — APPROVED** [review]

**Pull request merged**